### DISCUSSION TITLE
`@optional` support

### DISCUSSION THREAD

**Discussion by jrw972** [jmmorato/openddsharp#215] [General]
@jmmorato @tmayoff has started work on `@optional` support in https://github.com/OpenDDS/OpenDDS/pull/4364.
We are trying to develop a [roadmap and plans](https://github.com/OpenDDS/OpenDDS/wiki/Proposal:-@optional-support) for how `@optional` can be supported in various language bindings.
Please share your thoughts on how `@optional` may be incorporated into openddssharp so that we can make decisions that are beneficial for everyone.

Thanks.

**Comment by jmmorato** [comment]
Optional support is already defined in the `IDL4 to C# Language Mapping` documentation:

_IDL elements annotated with `@optional` whose type T maps to a C# value type shall map to System.Nullable<T>. IDL types mapped to reference types shall remain unchanged.
All mapped optional elements shall be annotated with OptionalAttribute, a custom attribute within the Omg.Types namespace that is defined as follows:_

```csharp
namespace Omg.Types
{
    public class OptionalAttribute : System.Attribute
    {
        public OptionalAttribute()
        {
            IsOptional = true;
        }
        public bool IsOptional { get; set; }
    }
}
```

_The “all values constructor” (see Clause 7.2.4.3.1) of the mapped C# class shall follow the same mapping in the definition of the input parameters associated with optional members, annotating each optional parameter with Omg.Types.OptionalAttribute as well._

_NOTE—Version 1.1 of the .NET Standard introduced OptionalAttribute, an attribute in the System.Runtime.InteropServices namespace that may be used to annotate optional parameters; however, the .NET Standard OptionalAttribute may not be used to annotate class properties, such as those representing IDL optional members. To mitigate this limitation, this specification introduces an alternative OptionalAttribute within the Omg.Types namespace that may be used for both optional properties as well. For consistency and simplicity (to minimize name collisions within the declaration of a class generated from the definition of an IDL structure containing optional members) this specification uses the OptionalAttribute within the Omg.Types namespace to annotate both optional class properties and optional parameters in the “all values constructor.”_

Based on that, here a short list that came to my mind of the work expected to support the optional feature in OpenDDSharp:
- Migrate OpenDDSharp to the latest OpenDDS version. 
  - I foresee some breaking changes due to the `ConfigStore` modifications in OpenDDS.
- Ensure using the correct C++ bindings (11/17) when generating the OpenDDS project for the data types
- Ensure JSON writers and JSON readers support optional feature in OpenDDS and OpenDDSharp
  - The implementation of the binding using CDR serialization is planned for future work in OpenDDSharp, similar requirement would be needed to support optional feature
  - The implementation of the binding using `DynamicData` is planned for future work in OpenDDSharp, similar requirement would be needed to support optional feature
- Provide a new C# `OpenDDSharp.Omg.Types.OptionalAttribute`
- Modify `openddsharp_idl` code generator to support nullable value types as described in the mapping documentation
- Unit Test `optional` feature code generation and publication/subcription
- Update IDL documentation

**Reply by jrw972** [reply]
Thanks.  I recorded the action items on the roadmap document.
